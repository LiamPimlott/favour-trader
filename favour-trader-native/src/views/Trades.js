import React, { Component } from 'react';
import { FlatList, StyleSheet, View, Picker, Text, TextInput, TouchableOpacity } from 'react-native';
import { List, ListItem } from "react-native-elements";
import { StackNavigator } from 'react-navigation';
import AuthService from '../components/AuthService.js';
import axios from 'axios';

export default class Trades extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            source: 'active',
            trades: [],
            userId: 0,
        };
        this.authService = new AuthService();
    }

    componentDidMount() {
        this.mounted = true;
        if (this.mounted) {
            this.fetchTrades();
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    async fetchTrades() {
        const endpoint = `http://favour-trader.appspot.com/api/contracts/${this.state.source}`;

        if (this.authService.loggedIn() && this.mounted) {
            const profile = await this.authService.getProfile();

            this.authService.getToken()
                .then(token => {
                    const config = {
                        headers: {
                            Authorization: token,
                        },
                    };

                    axios.get(endpoint, config)
                        .then(res => res.data)
                        .then(contracts => {
                            if (this.mounted) {
                                this.setState({
                                    trades: contracts,
                                    userId: profile.id,
                                });
                            }
                        })
                        .catch((err) => {
                            console.log(err);
                        });
                })
        }
    }

    setSource = (itemValue) => {
        this.setState({
            source: itemValue,
        }, this.fetchTrades);
    }

    renderSeparator = () => {
        return (
            <View
                style={{
                    height: 1,
                    width: "86%",
                    backgroundColor: "#CED0CE",
                    marginLeft: "14%"
                }}
            />
        );
    };

    renderTrades() {
        return (
            <FlatList
                data={this.state.trades}
                renderItem={({ item }) => (
                    <ListItem
                        roundAvatar
                        title={`${item.offeror.name.first} â‡‹ ${item.offeree.name.first}`}
                        subtitle={`This trade is currently ${item.status.toLowerCase()}`}
                        avatar={{ uri: 'https://cdn.onlinewebfonts.com/svg/img_104784.png' }}
                    // Add onPress prop to navigate to trade when it exists
                    />
                )}
                keyExtractor={item => item._id}
                ItemSeparatorComponent={this.renderSeparator}
            />
        );
    }

    render() {
        const { trades } = this.state;
        const { navigate } = this.props.navigation;
        return (
            <View>
                <Picker selectedValue={this.state.source}
                    onValueChange={this.setSource}
                    style={styles.picker}>
                    <Picker.Item label="Active Trades" value="active" />
                    <Picker.Item label="Sent Trades" value="sent" />
                    <Picker.Item label="Received Trades" value="received" />
                </Picker>
                {
                    (trades !== null && trades instanceof Array && trades.length !== 0) ? (
                        <View>
                            <List style={styles.container}>
                                {this.renderTrades()}
                            </List>
                        </View>
                    ) : (
                            <Text> {`Looks like you don't have any ${this.state.source.toLowerCase()} trades yet.`} </Text>
                        )
                }
            </View>
        )
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#7851a9',
        alignItems: 'center',
        justifyContent: 'center',
    },
    buttonContainer: {
        width: 400
    },
    title: {
        fontWeight: 'bold',
        fontSize: 40
    },
    intro: {
        fontSize: 15
    }
});
