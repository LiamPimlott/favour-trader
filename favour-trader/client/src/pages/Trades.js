import React, {Component} from 'react';
import axios from 'axios'
import {Row, Col} from 'antd';
import TradeCard from '../components/TradeCard';
import { Link } from 'react-router-dom';

class Trades extends Component {
    constructor() {
        super();
        this.state = {
            trades: [],
            userId: 0,
        };
    }

    componentWillMount() {
        this.mounted = true;
    }

    componentDidMount() {
        const { authService } = this.props;
        const { match: { params } } = this.props;
        const endpoint = params.source ?
            `/api/contracts/${params.source}` :
            `/api/contracts/active`;
		
        if (authService.loggedIn()) {
            const profile = authService.getProfile();
            const config = {
                headers: {
                    Authorization: authService.getToken()
                },
            };

            axios.get(endpoint, config)
            .then(res => res.data)
            .then(contracts => {
                if (this.mounted) {
                    this.setState({
                        trades: contracts,
                        userId: profile.id,
                    });
                }
            })
            .catch( (err) => {
                console.log(err);
            });
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    renderTrades = () => {
        const { trades, userId } = this.state;
        const { match: { params } } = this.props;
        const source = params.source;
        return (
            <div style={{textAlign: "center"}}>
                {
                    (trades) ? (
                        <div className={'container'}>
                            <h1>Matched Traders</h1>
                            <Row gutter={16}>
                                {
                                    trades.map(trade => {
                                        return (
                                            <Col key={trade._id} xs={24} sm={24} md={12} lg={8}>
                                                <TradeCard
                                                    trade={trade}
                                                    source={source}
                                                    userId={userId}
                                                />
                                            </Col>
                                        )
                                    }, this)
                                }
                            </Row>
                        </div>
                    ) : (
                        ''
                    )
                }
            </div>
        );
    }

    renderNoTrades = () => {
        const { match: { params } } = this.props;
        const source = params.source;
        switch (source) {
            case 'active':
            default:
                return (
                    <div>
                        <h3>
                            Looks like you don't have any active trades right now.
                        </h3>
                        <Link to={'/'}>
                            Search your matches!
                        </Link>
                    </div>
                )
            case 'sent':
                return (
                    <div>
                        <h3>
                            Looks like you haven't sent any trades lately.
                        </h3>
                        <Link to={'/filterWants'}>
                            Find what you're looking for!
                        </Link>
                    </div>
                )
            case 'received':
                return (
                    <div>
                        <h3>
                            Looks like you haven't received any trades lately.
                        </h3>
                        <Link to='/filterHas'>
                            Share your skills!
                        </Link>
                    </div>
                )
        }

    }

    render() {
        return (
            <div className={'center-helper'}>
                {this.state.trades !== null && this.state.trades.length !== 0 ? this.renderTrades() : this.renderNoTrades()}
            </div>
        );
    }
}

export default Trades;