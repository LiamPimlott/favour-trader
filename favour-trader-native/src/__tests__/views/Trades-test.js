import 'react-native';
import React from 'react';
import { shallow, configure } from 'enzyme';
import Trades from '../../views/Trades.js';
import toJson from 'enzyme-to-json'; //added this line

const Adapter = require("enzyme-adapter-react-16");
configure({adapter : new Adapter()});

describe('Testing Trades component', () => {
    it('renders as expected', () => {
        const trades = {
            recipientId: '',
            recipientFirstName: 'Test User Name',
            recipientFirstName: '',
            requestableFavours: [],
            currentStep: 1,
            creator: {},
            offerableFavours: [],
            requestedFavours: (new Map()),
            offeredFavours: (new Map()),
            tradeMessage: '',
            failedAttempt: false,
        }
        const navigation = { navigate: jest.fn() };
        const wrapper = shallow(
            <Trades trades={trades} navigation={navigation}/>
        );
        expect(toJson(wrapper)).toMatchSnapshot(); //edited this line
    });
});