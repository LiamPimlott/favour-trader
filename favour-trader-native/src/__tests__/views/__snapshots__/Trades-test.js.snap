// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Trades component renders as expected 1`] = `
<View>
  <Picker
    mode="dialog"
    onValueChange={[Function]}
    selectedValue="active"
  >
    <PickerItem
      label="Active Trades"
      value="active"
    />
    <PickerItem
      label="Sent Trades"
      value="sent"
    />
    <PickerItem
      label="Received Trades"
      value="received"
    />
  </Picker>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
     
    Looks like you don't have any active trades yet.
     
  </Text>
</View>
`;
